File: kernel/memory/slab.c
// SLAB memory allocation

#include <menix/common.h>
#include <menix/memory/pm.h>
#include <menix/memory/slab.h>

#include <string.h>

static Slab slabs[8] = {0};

static void slab_new(Slab* slab, usize size)
{
	slab->lock = spin_new();
	// Allocate a new page for the head.
	slab->head = (void**)((usize)pm_alloc(1) + pm_get_phys_base());
	slab->ent_size = size;

	const usize offset = ALIGN_UP(sizeof(SlabHeader), size);
	const usize available_size = CONFIG_page_size - offset;

	SlabHeader* ptr = (SlabHeader*)slab->head;
	ptr->slab = slab;
	slab->head = (void**)((void*)slab->head + offset);

	void** arr = slab->head;
	const usize max = available_size / size - 1;
	const usize fact = size / sizeof(void*);

	for (usize i = 0; i < max; i++)
	{
		arr[i * fact] = &arr[(i + 1) * fact];
	}
	arr[max * fact] = NULL;
}

void slab_init(void)
{
	// Create slabs for common structure sizes to minimize overhead.
	slab_new(&slabs[0], 8);
	slab_new(&slabs[1], 16);
	slab_new(&slabs[2], 32);
	slab_new(&slabs[3], 64);
	slab_new(&slabs[4], 128);
	slab_new(&slabs[5], 256);
	slab_new(&slabs[6], 512);
	slab_new(&slabs[7], 1024);
}

static void* slab_do_alloc(Slab* slab)
{
	spin_acquire_force(&slab->lock);

	if (slab->head == NULL)
		slab_new(slab, slab->ent_size);

	void** old_free = slab->head;
	slab->head = *old_free;
	memset(old_free, 0, slab->ent_size);

	spin_free(&slab->lock);
	return old_free;
}

static void slab_do_free(Slab* slab, void* addr)
{
	spin_acquire_force(&slab->lock);

	if (addr == NULL)
		goto cleanup;

	void** new_head = addr;
	*new_head = slab->head;
	slab->head = new_head;

cleanup:
	spin_free(&slab->lock);
}

// Finds a suitable slab that can contain `size` bytes.
// Returns `NULL` if none can do so.
static inline Slab* slab_find_size(usize size)
{
	for (usize i = 0; i < ARRAY_SIZE(slabs); i++)
	{
		if (slabs[i].ent_size >= size)
			return &slabs[i];
	}
	return NULL;
}

void* slab_alloc(usize size)
{
	// Find a suitable slab.
	Slab* slab = slab_find_size(size);
	// If there is already a usable slab, do allocation on that.
	if (slab != NULL)
		return slab_do_alloc(slab);

	// Get how many pages have to be allocated in order to fit `size`.
	usize num_pages = ROUND_UP(size, CONFIG_page_size);
	// Allocate the pages plus an additional page for metadata.
	PhysAddr ret = pm_alloc(num_pages + 1);
	// If the allocation failed, return NULL.
	if (ret == 0)
		return NULL;

	ret = ret + (PhysAddr)pm_get_phys_base();
	// Write metadata into the first page.
	SlabInfo* info = (SlabInfo*)ret;
	info->num_pages = num_pages;
	info->size = size;
	// Skip the first page and return the next one.
	return (void*)(ret + CONFIG_page_size);
}

void* slab_realloc(void* old, usize new_bytes)
{
	// If we didn't get a previous size, just treat this like a regular allocation.
	if (old == NULL)
		return slab_alloc(new_bytes);

	// If the address is page aligned.
	if ((usize)old == ALIGN_DOWN((usize)old, CONFIG_page_size))
	{
		SlabInfo* info = (SlabInfo*)(old - CONFIG_page_size);
		if (ROUND_UP(info->size, CONFIG_page_size) == ROUND_UP(new_bytes, CONFIG_page_size))
		{
			info->size = new_bytes;
			return old;
		}
		void* new = slab_alloc(new_bytes);
		if (new == NULL)
			return NULL;

		// Copy the old data over to the new memory.
		if (info->size > new_bytes)
			memcpy(new, old, new_bytes);
		else
			memcpy(new, old, info->size);

		slab_free(old);
		return new;
	}

	// Filter out the meta data.
	SlabHeader* slab_header = (SlabHeader*)((usize)old & ~0xFFF);
	Slab* slab = slab_header->slab;

	if (new_bytes > slab->ent_size)
	{
		void* new_addr = slab_alloc(new_bytes);
		if (new_addr == NULL)
			return NULL;

		memcpy(new_addr, old, slab->ent_size);
		slab_do_free(slab, old);
		return new_addr;
	}

	return old;
}

void slab_free(void* addr)
{
	if (addr == NULL)
		return;

	// If the address is page aligned.
	if ((usize)addr == ALIGN_DOWN((usize)addr, CONFIG_page_size))
	{
		SlabInfo* info = (SlabInfo*)(addr - CONFIG_page_size);
		pm_free(((PhysAddr)info - (PhysAddr)pm_get_phys_base()), info->num_pages + 1);
		return;
	}

	SlabHeader* header = (SlabHeader*)(ALIGN_DOWN((usize)addr, CONFIG_page_size));
	slab_do_free(header->slab, addr);
}

File: kernel/src/generic/memory/slab.rs
// Slab allocator

use super::{
    VirtAddr,
    pmm::{AllocFlags, KernelAlloc, PageAllocator},
    virt::VmLevel,
};
use crate::{
    arch,
    generic::util::{align_down, align_up, spin_mutex::SpinMutex},
};
use core::{
    alloc::{GlobalAlloc, Layout},
    hint::{likely, unlikely},
    mem::size_of,
    ptr::null_mut,
};

#[derive(Debug)]
struct Slab {
    /// Size of one entry.
    ent_size: usize,
    head: SpinMutex<VirtAddr>,
}

#[repr(transparent)]
struct SlabHeader {
    slab: *const Slab,
}

struct SlabInfo {
    /// Amount of pages connected to this slab.
    num_pages: usize,
    /// Size of this slab.
    size: usize,
}

impl Slab {
    /// Creates a new, uninitialized slab.
    const fn new(size: usize) -> Self {
        Self {
            ent_size: size,
            head: SpinMutex::new(VirtAddr::null()),
        }
    }

    /// Initializes a slab.
    fn init(&self) {
        unsafe {
            // Calculate the amount of bytes we need to skip in order to be able to store a reference to the slab.
            let offset = align_up(size_of::<SlabHeader>(), self.ent_size);
            // That also means we need to deduct that amount here.
            let available_size = (arch::virt::get_page_size(VmLevel::L1)) - offset;

            // Allocate memory for this slab.
            let mem = KernelAlloc::alloc(1, AllocFlags::empty()).expect("Out of memory");
            let mut head = mem.as_hhdm::<*mut ()>();

            // Get a reference to the start of the buffer.
            let ptr = head as *mut SlabHeader;
            // In that first entry, record a pointer to the head.
            (*ptr).slab = &raw const *self;
            // Now save that start to the slab.
            head = (head).byte_add(offset);

            let arr = head;
            let max = available_size / self.ent_size - 1;
            let fact = self.ent_size / size_of::<*mut ()>();

            for i in 0..max {
                *arr.add(i * fact) = arr.add((i + 1) * fact) as *mut ();
            }
            *arr.add(max * fact) = null_mut();

            *self.head.lock() = head.into();
        }
    }

    fn alloc(&self) -> *mut u8 {
        // Initialize the slab if not done already.
        if likely(*self.head.lock() == VirtAddr::null()) {
            self.init();
        }

        let mut head = self.head.lock();
        let old_free = head.value() as *mut *mut ();

        unsafe {
            *head = (*old_free).into();
        }

        return old_free as *mut u8;
    }

    fn free(&self, addr: *mut u8) {
        if unlikely(addr.is_null()) {
            return;
        }

        let new_head = addr as *mut *mut ();
        let mut head = self.head.lock();

        unsafe {
            *new_head = head.value() as *mut ();
        }
        *head = new_head.into();
    }
}

#[inline]
fn find_size(size: usize) -> Option<&'static Slab> {
    ALLOCATOR.slabs.iter().find(|&slab| slab.ent_size >= size)
}

#[repr(C, align(4096))]
pub struct SlabAllocator {
    slabs: [Slab; 8],
}

// Register the slab allocator as the global allocator.
#[global_allocator]
pub static ALLOCATOR: SlabAllocator = SlabAllocator {
    slabs: [
        Slab::new(16),
        Slab::new(32),
        Slab::new(64),
        Slab::new(128),
        Slab::new(256),
        Slab::new(512),
        Slab::new(1024),
        Slab::new(2048),
    ],
};

unsafe impl GlobalAlloc for SlabAllocator {
    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
        // If there's nothing to allocate, don't.
        if unlikely(layout.size() == 0) {
            return null_mut();
        }

        // Find a suitable slab.
        let slab = find_size(layout.size());
        if let Some(s) = slab {
            // The allocation fits within our defined slabs.
            let result = s.alloc();
            debug_assert!(result as usize % layout.align() == 0);
            return result;
        }

        // The allocation won't fit within our defined slabs.
        // Get how many pages have to be allocated in order to fit `size`.
        let num_pages = align_up(layout.size(), arch::virt::get_page_size(VmLevel::L1))
            / arch::virt::get_page_size(VmLevel::L1);

        // Allocate the pages plus an additional page for metadata.
        match KernelAlloc::alloc(num_pages + 1, AllocFlags::empty()) {
            Ok(mem) => unsafe {
                // Convert the physical address to a pointer.
                let ret: *mut u8 = mem.as_hhdm();

                // Write metadata into the first page.
                let info = ret as *mut SlabInfo;
                (*info).num_pages = num_pages;
                (*info).size = layout.size();

                // Skip the metadata and return the next one.
                return ret.byte_add(arch::virt::get_page_size(VmLevel::L1));
            },
            Err(_) => return null_mut(),
        }
    }

    unsafe fn dealloc(&self, ptr: *mut u8, _layout: Layout) {
        if ptr.is_null() {
            return;
        }

        unsafe {
            if ptr as usize == align_down(ptr as usize, arch::virt::get_page_size(VmLevel::L1)) {
                let info = ptr.sub(arch::virt::get_page_size(VmLevel::L1)) as *mut SlabInfo;
                KernelAlloc::dealloc((VirtAddr::from(info)).as_hhdm().unwrap(), (*info).num_pages);
            } else {
                let header = align_down(ptr as usize, arch::virt::get_page_size(VmLevel::L1))
                    as *mut SlabHeader;
                (*(*header).slab).free(ptr);
            }
        }
    }
}
